CMS-Enterprise/sbom-harbor-ui
# docs: add Storybook support

## Summary

This pull request introduces Storybook to the project, a powerful tool for developing and showcasing components in an isolated environment. It provides an interactive canvas for testing different states of your components. The addition of Storybook will enhance component development and debugging, improve component reusability, and serve as a living documentation for the components.

### Added

- **Storybook Configuration:**

  - In `.storybook/main.ts`, the configuration file for Storybook, the stories' path is set to `"../src/**/*.stories.@(js|jsx|ts|tsx|mdx)"`. This allows for any files with the extension `.stories.js`, `.stories.jsx`, `.stories.ts`, `.stories.tsx`, or `.stories.mdx` in the `src` directory or its subdirectories will be considered as a Storybook story. The addons used in the project, including `@storybook/addon-links`, `@storybook/addon-essentials`, and `storybook-addon-performance`, are also specified.
  - In `.storybook/preview.ts`, the Storybook preview configuration file, a decorator for Material-UI's theme provider is added. This decorator wraps every story with the theme provider, ensuring that the Material-UI theme is available in all stories. The default parameters for actions and backgrounds are also set.
  - In `storybook/README.md`, docs providing guidance for debugging issues with Storybook and Yarn Plug'n'Play (PnP) are added.

- **Stories:**

  - `src/stories/Introduction.mdx`: This file is a Storybook document that serves as the introduction for the storybook. It is the first page users see when they open the Storybook. The file is written in MDX, a markdown format that supports JSX, and explains the purpose and usage of the Storybook for the project. It provides a high-level overview of what Storybook is and how it's used within the project. It also contains links to additional resources about Storybook for those who want to learn more.
  - Three `.stories.tsx` files defines a set of Storybook stories for a specific component. For each component, a default export containing metadata about the component and the stories are defined. 
    - `src/components/stories/Copyright.stories.tsx`: This file contains the Storybook stories for the `Copyright` component. This component is very simple, so only one story is defined, displaying the `Copyright` component as it appears by default. 
    - `src/components/stories/MuiAvatar.stories.tsx`: The stories in this file are for the `MuiAvatar` component, a Material-UI avatar component. Two stories are defined: a default story that shows the `MuiAvatar` with a hardcoded image, and a story "With text" which shows the `MuiAvatar` with a simple text.
    - `src/components/stories/UserAvatar.stories.tsx`: These stories are for the `UserAvatar` component, a more complex component that displays a user avatar. Two stories are defined: "Default" which shows the `UserAvatar` with a default image, and "With text" that represents the `UserAvatar` with a user's initial.
  - In addition to showing the the basic usage of these components and providing a visual test bed for their different states, these files serve as a template for creating and structuring stories in the future.

### Changed

- In `package.json`, Storybook-related dependencies are added, allowing for the use of Storybook in the project.
- In `.eslintrc.cjs`, the ESLint configuration file, rules are added/updated to enable linting of Storybook files.

### Deprecated

- n/a

### Removed

- n/a

### Fixed

- n/a

## How to test

1. Pull the changes from this branch.
2. Run `yarn` to install dependencies.
3. Run `yarn storybook` to start the Storybook server. This should open a new browser tab with the Storybook interface, where you can interact with the newly added stories.
4. Check the newly added README in the Storybook directory for any issues with Storybook and Yarn PnP.
5. Examine the ESLint output to ensure that there are no linting errors in the newly added stories.
6. Explore the three simple stories added for the basic components.

## Screenshots

- ![image](https://github.com/CMS-Enterprise/sbom-harbor-ui/assets/1915925/4dbd8e4a-86fe-4cd5-b13f-4c36d451cfa6)
- ![image](https://github.com/CMS-Enterprise/sbom-harbor-ui/assets/1915925/c468fdfa-6cec-4b5d-bee8-f40b848e6d39)
- ![image](https://github.com/CMS-Enterprise/sbom-harbor-ui/assets/1915925/bbb7fadd-9fa9-446a-8e52-e62e766e84b2)
